tests/e2e/set-ffi-env-e2e

```diff
@@ -154,140 +167,313 @@ while [ $# -gt 0 ]; do
+        info_message "$description already exists in $file_path"
+        return 0
+    fi
+
+    if [[ -n "$description" ]]; then
+        info_message "Appending $description to $file_path..."
+    fi
+
+    echo "$content" >> "$file_path"
+
+    if [[ -n "$description" ]]; then
+        info_message "$description appended successfully"
+    fi
+}
+
+# Utility function to display section headers
+section_header() {
+    local title="$1"
+    local separator="${2:-==============================}"
+
+    info_message "$title"
+    info_message "$separator"
+}
+
+# Utility functions for systemctl operations
+enable_services() {
+    local services=("$@")
+    for service in "${services[@]}"; do
+        info_message "Enabling service: $service"
+        systemctl enable "$service"
+    done
+}
+
+start_services() {
+    local services=("$@")
+    for service in "${services[@]}"; do
+        info_message "Starting service: $service"
+        systemctl start "$service"
+    done
+}
+
+restart_services() {
+    local services=("$@")
+    for service in "${services[@]}"; do
+        info_message "Restarting service: $service"
+        systemctl restart "$service"
+    done
+}
+
+# Validate environment prerequisites
+validate_environment() {
+    # Check root privileges
+    if [ "$EUID" -ne 0 ]; then
+        echo -e "[${RED} ERROR ${CLR}] Please run as root this script. It requires to set limits inside a container which is not allowed by root."
+        exit 1
+    fi
+
+    # Check for ostree system
+    if stat /run/ostree-booted > /dev/null 2>&1; then
+        section_header "Warning: script can not run on ostree image"
+        exit 0
+    fi
+}
+
+# Error handling wrapper for commands
+execute_with_logging() {
+    local description="$1"
+    shift
+    local command=("$@")
+
+    info_message "Executing: $description"
+    if "${command[@]}"; then
+        info_message "Success: $description"
+    else
+        local exit_code=$?
+        info_message "Error: $description failed with exit code $exit_code"
+        return $exit_code
+    fi
+}
+
+# Utility function to report script execution time
+report_execution_time() {
+    local start_time="$1"
+    local title="${2:-Running time for this script}"
+
+    # Capture the end time
+    local end_time
+    end_time=$(date +%s)
+
+    # Calculate the duration in seconds
+    local duration=$((end_time - start_time))
+
+    # Calculate days, hours, minutes and seconds
+    local days=$((duration / 86400))
+    local hours=$(( (duration % 86400) / 3600 ))
+    local minutes=$(( (duration % 3600) / 60 ))
+    local seconds=$((duration % 60))
+
+    info_message "${GRN}${title}${CLR}"
+    info_message "\t- ${days} days, ${hours} hours, ${minutes} minutes and ${seconds} seconds"
+}
+
+# Check if QM is already installed and active
+check_qm_installation_status() {
+    if ! rpm -q qm &>/dev/null; then
+        info_message "Warning: QM package not found, proceeding with installation"
```

Comment:
Isn't this an error and the tests shouldn't proceed?

Reply:
I don't think so because the next function will be setup_qm_services. setup_qm_services will create the RPMs and install qm, so I think it's fine as it is

Reply:
Ah ok. In that context it makes sense. These bash scripts for testing aren't really nice to read.